File: 2023/lib/2023/9.ex
import AOC
import AOCHelpers

aoc 2023, 9 do
  def p1(input) do
    input
    |> lines_of_integers()
    |> Enum.map(&next_in_sequence/1)
    |> Enum.sum()
  end

  def p2(_input) do
  end

  def next_in_sequence(xs) do
    xs
    |> differentiate()
    |> Enum.map(&List.last/1)
    |> Enum.sum()
  end

  def differentiate(xs) do
    xs
    |> Stream.iterate(&deltas/1)
    |> Enum.take_while(&(not all_zeros?(&1)))
  end

  def deltas(xs), do: Enum.zip_with(xs, Enum.drop(xs, 1), &(&2 - &1))

  def all_zeros?(xs), do: Enum.all?(xs, is?(0))
end

File: 2023/9.hs
#!/usr/bin/env runhaskell

import Control.Monad (ap)

parseInput :: String -> [[Integer]]
parseInput = map (map read . words) . lines

solve str = show (part1', part2')
  where
    input = parseInput str
    part1' = part1 input
    part2' = part2 input

part1 = sum . map next
part2 = sum . map prev

deltas = ap (zipWith subtract) tail
differentiate = takeWhile notAllZeros . iterate deltas
  where
    notAllZeros = not . all (0 ==)

next = sum . map last . differentiate

prev = foldl1 subtract . reverse . map head . differentiate

main :: IO ()
main = do
  contents <- getContents
  putStrLn $ solve contents
